backend/remote-state/etcdv3/client.go

```diff
@@ -0,0 +1,195 @@
+	etcdv3sync "github.com/coreos/etcd/clientv3/concurrency"
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/terraform/state"
+	"github.com/hashicorp/terraform/state/remote"
+)
+
+const (
+	lockAcquireTimeout = 2 * time.Second
+	lockInfoSuffix     = ".lockinfo"
+	lockSuffix         = ".lock"
+)
+
+// RemoteClient is a remote client that will store data in etcd.
+type RemoteClient struct {
+	Client *etcdv3.Client
+	DoLock bool
+	Key    string
+
+	etcdMutex   *etcdv3sync.Mutex
+	etcdSession *etcdv3sync.Session
+	info        *state.LockInfo
+	mu          sync.Mutex
+	modRevision int64
+}
+
+func (c *RemoteClient) Get() (*remote.Payload, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	res, err := c.Client.KV.Get(context.TODO(), c.Key)
+	if err != nil {
+		return nil, err
+	}
+	if res.Count == 0 {
+		return nil, nil
+	}
+	if res.Count >= 2 {
+		return nil, fmt.Errorf("Expected a single result but got %d.", res.Count)
+	}
+
+	c.modRevision = res.Kvs[0].ModRevision
+
+	payload := res.Kvs[0].Value
+	md5 := md5.Sum(payload)
+
+	return &remote.Payload{
+		Data: payload,
+		MD5:  md5[:],
+	}, nil
+}
+
+func (c *RemoteClient) Put(data []byte) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	res, err := etcdv3.NewKV(c.Client).Txn(context.TODO()).If(
+		etcdv3.Compare(etcdv3.ModRevision(c.Key), "=", c.modRevision),
+	).Then(
+		etcdv3.OpPut(c.Key, string(data)),
+		etcdv3.OpGet(c.Key),
+	).Commit()
+
+	if err != nil {
+		return err
+	}
+	if !res.Succeeded {
+		return fmt.Errorf("The transaction did not succeed.")
+	}
+	if len(res.Responses) != 2 {
+		return fmt.Errorf("Expected two responses but got %d.", len(res.Responses))
+	}
+
+	c.modRevision = res.Responses[1].GetResponseRange().Kvs[0].ModRevision
+	return nil
+}
+
+func (c *RemoteClient) Delete() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	_, err := c.Client.KV.Delete(context.TODO(), c.Key)
+	return err
+}
+
+func (c *RemoteClient) Lock(info *state.LockInfo) (string, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if !c.DoLock {
+		return "", nil
+	}
+
+	c.info = info
+	return c.lock()
+}
+
+func (c *RemoteClient) Unlock(id string) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if !c.DoLock {
+		return nil
+	}
+
+	return c.unlock(id)
+}
+
+func (c *RemoteClient) getLockInfo() (*state.LockInfo, error) {
+	res, err := c.Client.KV.Get(context.TODO(), c.Key+lockInfoSuffix)
+	if err != nil {
+		return nil, err
+	}
+	if res.Count == 0 {
+		return nil, nil
+	}
+
+	li := &state.LockInfo{}
+	err = json.Unmarshal(res.Kvs[0].Value, li)
+	if err != nil {
+		return nil, fmt.Errorf("Error unmarshaling lock info: %s.", err)
+	}
+
+	return li, nil
+}
+
+func (c *RemoteClient) putLockInfo(info *state.LockInfo) error {
+	c.info.Path = c.Key
+	c.info.Created = time.Now().UTC()
+
+	_, err := c.Client.KV.Put(context.TODO(), c.Key+lockInfoSuffix, string(c.info.Marshal()))
+	return err
+}
+
+func (c *RemoteClient) lock() (string, error) {
+	session, err := etcdv3sync.NewSession(c.Client)
```

Comment:
I'm not sure what etcd's behavior would be here, but terraform should never recursively lock, so you can check for an existing session and return an error if there is already a lock outstanding.

Reply:
Done 👍  9aa417e